lexicon/modification.py

```diff
@@ -0,0 +1,156 @@
+        else:
+            lex.phonemes = self.static_lexicon.phonemes
+
+        if self.sort_lemmas:
+            lemma_dict = {}
+            for lemma in self.static_lexicon.lemmas:
+                # sort by first orth entry
+                lemma_dict[lemma.orth[0]] = lemma
+            lex.lemmas = [lemma_dict[key] for key in sorted(lemma_dict.keys())]
+        else:
+            lex.lemmas = self.static_lexicon.lemmas
+
+        write_xml(self.out_bliss_lexicon.get_path(), lex.to_xml())
+
+
+class MergeLexiconJob(Job):
+    """
+    Merge multiple bliss lexica into a single bliss lexicon.
+
+    Phonemes and lemmas can be individually sorted alphabetically or kept as is.
+
+    When merging a lexicon with a static lexicon, putting the static lexicon first
+    and only sorting the phonemes will result in the "typical" lexicon structure.
+    """
+
+    def __init__(self, bliss_lexica, sort_phonemes, sort_lemmas, compressed=False):
+        """
+        :param list[Path] bliss_lexica: list of bliss lexicon files (plain or gz)
+        :param bool sort_phonemes: sort phoneme inventory alphabetically
+        :param bool sort_lemmas: sort lemmas alphabetically based on first orth entry
+        :param bool compressed: compress final lexicon
+        """
+        self.lexica = bliss_lexica
+        self.sort_phonemes = sort_phonemes
+        self.sort_lemmas = sort_lemmas
+
+        self.out_bliss_lexicon = self.output_path(
+            "lexicon.xml.gz" if compressed else "lexicon.xml"
+        )
+
+    def tasks(self):
+        yield Task("run", mini_task=True)
+
+    def run(self):
+        merged_lex = lexicon.Lexicon()
+
+        lexica = []
+        for lexicon_path in self.lexica:
+            lex = lexicon.Lexicon()
+            lex.load(lexicon_path.get_path())
+            lexica.append(lex)
+
+        # combine the phonemes
+        merged_phonemes = OrderedDict()
+        for lex in lexica:
+            for symbol, variation in lex.phonemes.items():
+                if symbol in merged_phonemes.keys():
+                    assert variation == merged_phonemes[symbol], (
+                        "conflicting phoneme variant for phoneme: %s" % symbol
+                    )
+                else:
+                    merged_phonemes[symbol] = variation
+
+        if self.sort_phonemes:
+            sorted_phoneme_list = [
+                (k, merged_phonemes[k]) for k in sorted(merged_phonemes.keys())
+            ]
+            for phoneme_tuple in sorted_phoneme_list:
+                merged_lex.add_phoneme(
+                    symbol=phoneme_tuple[0], variation=phoneme_tuple[1]
+                )
+        else:
+            merged_lex.phonemes = merged_phonemes
+
+        # combine the lemmas
+        if self.sort_lemmas:
+            lemma_dict = {}
+            for lex in lexica:
+                for lemma in lex.lemmas:
+                    # sort by first orth entry
+                    lemma_dict[lemma.orth[0]] = lemma
```

Comment:
If one first orth exists in two lexica the old lemma is overwritten. Would we want/expect that?

Reply:
No, this is inconsistent with the behavior of not sorting lemmas. There no merging takes place. I.e. here we should make lemma_dict a `str -> list[Lemma]` dict.

Reply:
Yes, this is inconsistent, but the question is how would we want that? I would just add an assert for now that crashes the job if there is any overlap.

Reply:
I'd just keep the two lemmas as is and include them both. Just like in the non-sorted case.

Reply:
But can this ever be the desired result? So how will RASR treat two lemmata with the same orth?

Reply:
It will emit a warning saying that it will use the first lemma when mapping orth -> lemma. But in decoding this is not a problem.

Reply:
Also what do you do when the orth is empty? E.g. for silence? Sentence-begin/Sentence-end might also have empty orth. This needs to be considered here (i.e. you cannot access lemma.orth[0] without checking for the length of orth).

Reply:
An empty orth is a zero length string, so the list will still contain one entry

Reply:
> It will emit a warning saying that it will use the first lemma when mapping orth -> lemma. But in decoding this is not a problem.

But if it uses the first, we would not want to have a lemma in there that is never used, or not? I would rather go for the version that it crashes when there is a duplicate, and we later add an option that allows merging if the actually get the case.

Reply:
But you already now might have multiple lemmas with empty orth (i.e. silence, sentence-end, sentence-begin)

Reply:
Or you might have lemmas without any orth.